plugins/kafka/kafka.cpp

```diff
@@ -891,23 +890,29 @@ namespace KafkaPlugin
         private:
 
             std::atomic_bool    shouldRun;      //!< If true, we should execute our thread's main event loop
-    } *publisherCacheExpirer;
+    };
+
+    //----------------------------------
+
+    static Singleton<PublisherCacheExpirerObj> publisherCacheExpirer;
+
+    static PublisherCacheExpirerObj & queryPublisherCacheExpirer()
+    {
+        return *publisherCacheExpirer.query([] () { return new PublisherCacheExpirerObj; });
+    }
 
     //--------------------------------------------------------------------------
-    // Lazy Initialization
+    // Singleton Initialization
     //--------------------------------------------------------------------------
 
     /**
      * Make sure the publisher object cache is initialized as well as the
-     * associated background thread for expiring idle publishers.  This is
-     * called only once.
+     * associated background thread for expiring idle publishers.
      */
     static void setupPublisherCache()
     {
-        KafkaPlugin::publisherCache = new KafkaPlugin::PublisherCacheObj();
-
-        KafkaPlugin::publisherCacheExpirer = new KafkaPlugin::PublisherCacheExpirerObj;
-        KafkaPlugin::publisherCacheExpirer->start();
+        queryPublisherCache();
+        queryPublisherCacheExpirer().start();
```

Comment:
Potential race condition - two threads could get here at the same time, and both pass the !isAlive() test.
I think the correct solution (using an auto pattern jake uses) is 

```
 static PublisherCacheExpirerObj & queryPublisherCache()
 {
     auto initFunction = [] ()
    {
        PublisherCacheExpirerObj * publisher = new PublisherCacheExpirerObj;
        KafkaPlugin::publisherCacheExpirer = new KafkaPlugin::PublisherCacheExpirerObj;
        KafkaPlugin::publisherCacheExpirer->start();
        return publisher;
    };
    return *publisherCacheExpirer.query(initFunction);
}
```

Reply:
Then you can delete the calls to setupPublisherCache() since queryPublisherCache() will handle it for you.

Reply:
Due to the circular dependencies between PublisherCacheObj and PublisherCacheExpirerObj I opted to refactor:  The expirer class (and its background thread) are now gone.  Expiration is done after a new publisher object is created.  This simplifies things a great deal and probably increases system performance slightly (one less thread to manage).